File: python/bin/pipe.py
#!/usr/bin/python2

import base64
import json
import sys
import traceback

# TODO import the algorithm apply method so this doesn't crash

FIFO_PATH = '/tmp/algoout'

def main():
    for line in sys.stdin:
        request = json.loads(line)
        response = get_response(request)

        # Add final newline delimeter and flush stdout before writing back response
        sys.stdout.write('\n')
        sys.stdout.flush()

        with open(FIFO_PATH, 'w') as f:
            f.write(json.dumps(response))
            f.write('\n')

def get_response(request):
    try:
        result = call_algorithm(request)
        if isinstance(result, bytearray):
            content_type = 'binary'
        else:
            content_type = 'json'

        response = {
            'result': result,
            'metadata': {
                'content_type': content_type
            }
        }
    except Exception as e:
        _, _, exc_traceback = sys.exc_info()

        response = {
            'message': str(e),
            'stacktrace': traceback.format_exc(exc_traceback)
        }

    return response

def call_algorithm(request):
    if request['content_type'] in ['text', 'json']:
        data = request['data']
    elif request['content_type'] == 'binary':
        data = base64.b64encode(request['data'])
    else:
        raise Exception("Invalid content_type: {}".format(request['content_type']))

    return apply(data)

if __name__ == '__main__':
    main()

File: ruby/bin/pipe.rb
require 'json'
require 'base64'
require_relative "#{Dir.pwd}/src/algorithm.rb" # TODO: would be nice to know algoname

def pipe_loop
  STDIN.each_line do |line|
    request = JSON.parse(line)
    response = begin
      result = call_algorithm(request)
      content_type = if result.is_a? String then
        result.encoding == 'ASCII-8BIT' ? 'binary' : 'text'
      else
        'json'
      end
      { :result => result, :metadata => { :content_type => content_type}}
    rescue Exception => e
      { :error => { :message => e.message, :stacktrace => e.backtrace.join("\n") }}
    end

    # Add final newline delimeter and flush stdout before writing back response
    puts "\n"
    STDOUT.flush
    File.open('/tmp/algoout', 'w') do |algoout|
      algoout.puts response.to_json
    end
  end
end

def call_algorithm(request)
  input = case request['content_type']
    when 'text', 'json'
      request['data']
    when 'binary'
      Base64.decode64(request['data'])
    else
      raise "Invalid content_type: #{request['content_type']}"
  end
  apply(input)
end

pipe_loop
